### PR TITLE
Support per-parameter test decoration

### LINKED ISSUE #79161: Support more granular test decoration
Currently, the `DecorateInfo` class only supports conditionally applying decorators to test functions based on test class name, base test name, device, and dtype. Example usage:

https://github.com/pytorch/pytorch/blob/7cb4a76844edfed7e31c50ec5af8fbfe8cbb6ca0/torch/testing/_internal/common_methods_invocations.py#L9800-L9805

It would be useful to support more granular test decoration, ideally based on any parameter that is passed to the test function rather than just `device` / `dtype`.

Case in point: #78735 is adding support for running tests across train / eval modes for `ModuleInfo`. Some tests are expected to fail for eval mode only, and there's no granular way to indicate this.

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #79979
* #80471

Fixes #79161

This PR does the following:
* Expands the `parametrize_fn()` signature from returning a 3-tuple of `(test, test_name, param_kwargs)` to returning a 4-tuple of `(test, test_name, param_kwargs, decorator_fn)`. Expected signature for the addition is `decorator_fn(param_kwargs) -> List[decorator]` i.e. given the full set of test params, return a list of decorators to apply.
    * `@modules`, `@ops`, and `@parametrize` now fit the new signature, returning `decorator_fn`s instead of applying decorators themselves.
    * `instantiate_parametrized_tests()` and `instantiate_device_type_tests()` now call the returned `decorator_fn`, passing in the full set of `param_kwargs` (after composition + `device` / `dtype` additions) and applying the returned decorators.
    * Composing multiple `parametrize_fn`s also composes the corresponding `decorator_fn`s; the composed `decorator_fn` simply concatenates the decorator lists returned by the constituents.
* Expands `DecorateInfo.is_active` to support callables:
```python
DecorateInfo(
    unittest.expectedFailure, "TestOps", "test_python_ref_executor",
    device_type='cuda', active_if=lambda params: params['executor'] == 'nvfuser'
),
```
* Adds several tests to `test/test_testing.py` ensuring proper decoration using `@parametrize` and `@modules`.
* (minor) Refactors `ModuleInfo` slightly to better match `OpInfo`; replace `should_skip()` with `get_decorators()` and merge `skips` and `decorators`.
* (minor) Fixes a couple `ModuleInfo` naming oddities uncovered during testing.

### CODE REVIEW

File: torch/testing/_internal/common_device_type.py [hunk 1]
```diff
@@ -355,19 +363,23 @@ def _apply_precision_override_for_test(self, test, param_kwargs):
...
-        def instantiate_test_helper(cls, name, *, test, param_kwargs=None):
+        def instantiate_test_helper(cls, name, *, test, param_kwargs=None, decorator_fn=lambda _: []):
+            # Add the device param kwarg if the test needs device or devices.
+            param_kwargs = {} if param_kwargs is None else param_kwargs
+            test_sig_params = inspect.signature(test).parameters
+            if 'device' in test_sig_params or 'devices' in test_sig_params:
+                device_arg: str = cls.get_primary_device()
+                if hasattr(test, 'num_required_devices'):
+                    device_arg = cls.get_all_devices()
+                _update_param_kwargs(param_kwargs, 'device', device_arg)
```
Comment (jbschlosser) [inline torch/testing/_internal/common_device_type.py:374]: This was moved here from below so that the `device` kwarg is present in `param_kwargs` when passed to the `decorator_fn()`.

File: torch/testing/_internal/common_device_type.py [hunk 2]
```diff
@@ -312,6 +320,17 @@ def rel_tol(self, prec):
...
+    @classmethod
+    def _init_and_get_primary_device(cls):
+        try:
+            return cls.get_primary_device()
+        except Exception:
+            # For CUDATestBase, XLATestBase, and possibly others, the primary device won't be available
+            # until setUpClass() sets it. Call that manually here if needed.
+            if hasattr(cls, 'setUpClass'):
+                cls.setUpClass()
+            return cls.get_primary_device()
```
Comment (jbschlosser) [inline torch/testing/_internal/common_device_type.py:332]: This is a bit messy. The device specific test class for CUDA, XLA, and maybe others (?) define `setupUpClass()` method that populates the primary device. In the past, this has been run by `unittest` just before the first test invocation. Unfortunately, that's too late if we want the per-param decoration mechanism to have access to it during test generation time.

The hacky approach here calls `setUpClass()` if trying to get the primary device fails. Open to better suggestions! Note that I have been operating under the assumption that there may be other device-specific test classes out there we don't know about or can't easily change. If we aren't restricted by this, other options involving changing `CUDATestBase` / `XLATestBase` open up.

File: test/test_testing.py [hunk 3]
```diff
@@ -1494,6 +1500,48 @@ def test_two_things_custom_names(self, x, y):
         test_names = _get_test_names_for_test_class(TestParametrized)
         self.assertEqual(expected_test_names, test_names)
 
+    def test_apply_param_specific_decorators(self):
```
Comment (mruberry) [inline test/test_testing.py:1503]: Nice tests and comments

**Review by mruberry — APPROVED** [review]
Stamped!

I triggered some additional jobs, too

**Comment by mruberry** [comment]
Is there some documentation this should update? Maybe a comment like the one describing the OpInfos or the device-generic test framework?

File: torch/testing/_internal/common_device_type.py (hunk 2 above)
Comment (mruberry) [reply torch/testing/_internal/common_device_type.py:332]: I also can't immediately think of a better way to do this; I found the comment very helpful
